docs/site/Dynamic-models-repositories-controllers.md

```diff
@@ -0,0 +1,218 @@
+on `RepositoryMixin` apps.
+
+### Defining a Repository
+
+Now that a datasource is ready, a repository can be created for `BookModel`.
+
+The `@loopback/rest-crud` package exports a helper function,
+`defineCrudRepositoryClass`, this function creates a
+`defineCrudRepositoryClass`-based entity CRUD repository class for the given
+model.
+
+```ts
+const BookRepository = defineCrudRepositoryClass(BookModel);
+```
+
+If you want to customize the repository, you can create a copy of
+`defineCrudRepositoryClass`'s
+[implementation](https://github.com/strongloop/loopback-next/blob/00917f5a06ea8a51e1f452f228a6b0b7314809be/packages/rest-crud/src/repository-builder.ts#L25-L47)
+and modify it according to your requirements.
+
+If you want to use a non-`defineCrudRepositoryClass` repository class as the
+base for your repository, you can use the `defineRepositoryClass()` function
+instead.
+
+```ts
+import defineRepositoryClass from '@loopback/repository';
+// Assuming UserRepository is a pre-existing Repository class in the app
+const StudentRepository = defineRepositoryClass(BookModel, UserRepository);
+```
+
+Dependency injection has to be manually done for the datasource as shown below.
+
+```ts
+inject(`datasources.${dsName.name}`)(BookRepository, undefined, 0);
+const repoBinding = app.repository(BookRepository);
+```
+
+Note, the `app.repository()` method will be available only on `RepositoryMixin`
+apps.
+
+### Defining a Controller
+
+Once a repository is set up, a controller can be created for the model using
+the `defineCrudRestController` helper function from the `@loopback/rest-crud`
+package. It accepts a Model class and a `CrudRestControllerOptions` object.
+Dependency injection for the controlller has to be manually done.
+
+```ts
+const basePath = '/' + bookDef.name;
+const BookController = defineCrudRestController(BookModel, {basePath});
+inject(repoBinding.key)(BookController, undefined, 0);
+```
+
+The controller is then attached to the app by calling the `app.controller()`
+method.
+
+```ts
+app.controller(BookController);
+```
+
+The new CRUD REST endpoints for the model will be available on the app now.
+
+If you want a customized controller, you can create a copy of
+`defineCrudRestController`'s
+[implementation](https://github.com/strongloop/loopback-next/blob/00917f5a06ea8a51e1f452f228a6b0b7314809be/packages/rest-crud/src/crud-rest.controller.ts#L129-L269)
+and modify it according to your requirements.
+
+For details about `defineCrudRestController` and `CrudRestControllerOptions`,
+refer to the
+[@loopback/rest-crud API documentation](https://loopback.io/doc/en/lb4/apidocs.rest-crud.html).
+
+## Model discovery
+
+Some datasource connectors provide methods for discovering model definitions
+from existing database schema. The following APIs can be very useful when
+defining LoopBack models dynamically.
+
+```ts
+// List database tables and/or views
+ds.discoverModelDefinitions({views: true, limit: 20}, cb);
```

Comment:
Uh oh, isn't there a Promise based version of this method? All public API in LB4 is based on Promises, we don't use callbacks anywhere. If discovery-related DataSource methods are not Promise-enabled yet, then please open a follow-up issue to improve that and update this part of our docs too.

Having said that, AFAICT, `discoverModelDefinitions` does implement support for Promise style invocation and the typings describe both callback and promise styles.

Please:
1. Check which discovery-related methods support Promises and update examples here in docs to use `await`.
2. If there are any methods you would like to show here but that don't support Promises, then open a follow-up issue to fix that.

```suggestion
await ds.discoverModelDefinitions({views: true, limit: 20});
```